synapse/storage/databases/main/event_federation.py

```diff
@@ -207,6 +219,18 @@ def _get_auth_chain_difference_using_cover_index_txn(
                 seen_chains.add(chain_id)
                 chain_to_event.setdefault(chain_id, {})[sequence_number] = event_id
 
+        # Check that we actually have a chain ID for all the events.
+        events_missing_chain_info = initial_events.difference(chain_info)
+        if events_missing_chain_info:
+            # This can happen due to e.g. downgrade/upgrade of the server. We
+            # raise an exception and fall back to the previous algorithm.
+            logger.info(
+                "Unexpectedly found that events don't have chain IDs in room %s: %s",
+                room_id,
+                events_missing_chain_info,
+            )
```

Comment:
Is there any way to recover from this? Will it happen automatically?

Reply:
It should mostly happen automatically, unless there are events we don't have the full auth chain for (I have a couple of those on jki.re somehow). There's not really a way to recover from that failure mode, so I'm tempted to just leave it as is and revisit if it turns out that it happens a lot.